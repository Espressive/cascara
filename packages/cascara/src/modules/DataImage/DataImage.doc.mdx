---
title: DataImage
propTable: ./DataImage.js
type: 'module'
public: false
---

A data module to display an input **image** field

### Sandbox

Feel free to play around with the following code:

```jsx
<ModuleSandbox
  isEditing={true}
  record={{
    id: 1,
    firstName: 'John',
    lastName: 'Doe',
  }}
>
  <DataImage
    attribute='firstName'
    alt='My Text'
    src='/media/examples/login-button.png'
  />
</ModuleSandbox>
```
