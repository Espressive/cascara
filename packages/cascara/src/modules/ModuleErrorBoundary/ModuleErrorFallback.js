import React from 'react';

import styles from '../DataModule.module.scss';

const ModuleErrorFallback = () => {
  return (
    <div className={styles.Error} role='alert'>
      Error
    </div>
  );
};

export default ModuleErrorFallback;
