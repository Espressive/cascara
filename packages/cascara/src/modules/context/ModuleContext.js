import { createContext } from 'react';

// We should ONLY put values here that are needed for ALL Modules
const defaultValue = {
  isEditing: false,
  isEditingAny: false,
  value: 'blue',
};

/*
 * We will use this baseline context to create any of our context providers from.
 * Any unique functionality should be included in that provider and not in this Context.
 */
const ModuleContext = createContext(defaultValue);

/*
 * We are also exporting the default value in case we need to hydrate or merge
 * any providers with these default values.
 */
export { defaultValue };

export default ModuleContext;
