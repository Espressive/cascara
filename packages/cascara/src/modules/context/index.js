export { default as ModuleContext } from './ModuleContext';
export { default as ModuleProvider } from './ModuleProvider';
