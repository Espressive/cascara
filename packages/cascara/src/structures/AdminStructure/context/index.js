export { default as AdminContext } from './AdminContext';
export { default as AdminProvider } from './AdminProvider';
