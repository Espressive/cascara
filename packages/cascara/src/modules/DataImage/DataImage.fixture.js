import React from 'react';
import DataImage from './DataImage';
import ModuleSandbox from '../ModuleSandbox';

const DataImageSandbox = ({ isEditing, ...rest }) => (
  <ModuleSandbox isEditing={isEditing}>
    <DataImage {...rest} />
  </ModuleSandbox>
);

const srcPath = '/media/examples/my-button.png';

const displayProps = {
  label: 'Display',
  src: srcPath,
};

const editingProps = {
  isEditing: true,
  label: 'Editing',
  src: srcPath,
  alt: 'Start',
};

// These can be used in tests
export { displayProps, editingProps };

// These are our fixtures
export default {
  display: <DataImageSandbox {...displayProps} />,
  displayNoLabel: <DataImageSandbox {...displayProps} isLabeled={false} />,
  editing: <DataImageSandbox {...editingProps} />,
  editingNoLabel: <DataImageSandbox {...editingProps} isLabeled={false} />,
};
