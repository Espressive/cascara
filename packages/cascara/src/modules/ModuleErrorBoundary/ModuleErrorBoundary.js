import React from 'react';
import pt from 'prop-types';

import { Boundaries } from '../../system-components';
import ModuleErrorFallback from './ModuleErrorFallback';

const propTypes = {
  children: pt.oneOfType([pt.arrayOf(pt.node), pt.node]),
};

const ModuleErrorBoundary = ({ children }) => {
  return (
    <Boundaries ErrorComponent={ModuleErrorFallback}>{children}</Boundaries>
  );
};

ModuleErrorBoundary.propTypes = propTypes;

export default ModuleErrorBoundary;
