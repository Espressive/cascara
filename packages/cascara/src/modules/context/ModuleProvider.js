import React from 'react';
import pt from 'prop-types';

import ModuleContext, { defaultValue } from './ModuleContext';
import useToggle from '../../hooks/useToggle';

const propTypes = {
  children: pt.oneOfType([pt.element, pt.arrayOf(pt.element)]),
  value: pt.shape({
    isEditing: pt.bool,
  }),
};

const ModuleProvider = ({ children, value, ...props }) => {
  const [isEditing, setIsEditing] = useToggle(value?.isEditing || false);

  const mergedValues = {
    ...defaultValue,
    isEditing,
    setIsEditing,
    ...value,
  };

  return (
    <ModuleContext.Provider value={mergedValues} {...props}>
      {children}
    </ModuleContext.Provider>
  );
};

ModuleProvider.propTypes = propTypes;

export default ModuleProvider;
