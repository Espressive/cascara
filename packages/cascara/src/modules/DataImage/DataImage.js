import React, { useContext } from 'react';
import { Input } from 'reakit/Input';
import pt from 'prop-types';
import { ModuleContext } from '../context';
import styles from '../DataModule.module.scss';

import ModuleErrorBoundary from '../ModuleErrorBoundary';
import getAccessibleLabelSetters from '../helpers';

const propTypes = {
  /** A module can have an Attribute, which will be used as form field name */
  attribute: pt.string,
  /** A Module can be defined to not present an editing state */
  isEditable: pt.bool,
  /** Presents the input without a label. NOT USER CONFIGURABLE */
  isLabeled: pt.bool,
  /** A Module needs to have a unique label relative to its context */
  label: pt.string,
  /** Image path */
  src: pt.string,
  /** A Module can have a value */
  value: pt.string,
};

const DataImage = ({
  attribute,
  isEditable = true,
  isLabeled = true,
  label,
  value,
  src,
  ...rest
}) => {
  const { isEditing, formMethods } = useContext(ModuleContext);
  const { setAriaLabel, setHtmlFor } = getAccessibleLabelSetters(
    isLabeled,
    label
  );

  const renderEditing = (
    <label htmlFor={setHtmlFor}>
      {label && isLabeled && <span className={styles.LabelText}>{label}</span>}
      <Input
        {...rest}
        aria-label={setAriaLabel}
        className={styles.Image}
        defaultValue={value}
        id={label}
        name={attribute || label}
        ref={formMethods?.register}
        src={src}
        type={'image'}
      />
    </label>
  );

  const renderDisplay = (
    <span>
      {label && isLabeled && <span className={styles.LabelText}>{label}</span>}
      <span aria-label={label} className={styles.Image} {...rest}>
        {src}
      </span>
    </span>
  );

  // Do not render an editable input if the module is not editable
  return (
    <ModuleErrorBoundary>
      <div className={styles.Image}>
        {isEditing && isEditable ? renderEditing : renderDisplay}
      </div>
    </ModuleErrorBoundary>
  );
};

DataImage.propTypes = propTypes;

export { propTypes };
export default DataImage;
