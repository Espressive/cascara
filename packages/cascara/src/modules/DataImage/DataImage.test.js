import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import DataImage from './DataImage';
import { DisplayDataImage } from './DataImage.fixture';

describe('Input Image', () => {
  // without ModuleSandbox will render the property information into a span
  describe('default', () => {
    const testId = 'default';

    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(<DataImage data-testid={testId} />).container;
    });

    test('renders without any props', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <span> by default', () => {
      const input = screen.getByTestId(testId);
      // Make sure the actual DOM element is not render an input
      expect(input.tagName).toMatch('SPAN');
    });
  });

  describe('editing', () => {
    // specifying ModuleSandbox will have the flag to render our input date field
    const testId = 'default';
    beforeEach(() =>
      render(
        <DisplayDataImage
          alt='Login'
          data-testid={testId}
          isEditing
          src='/media/examples/my-button.png'
        />
      )
    );
    test('renders a <input date> by default', () => {
      const input = screen.getByTestId(testId);
      // Check that we also use the correct type for accessibility
      expect(input).toHaveAttribute('type', 'image');
    });

    test('renders the src', () => {
      const input = screen.getByTestId(testId);
      expect(input).toHaveAttribute('alt', 'Login');
      expect(input).toHaveAttribute('src', '/media/examples/my-button.png');
    });
  });
});
