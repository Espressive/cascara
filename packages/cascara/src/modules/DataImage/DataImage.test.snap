// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataDateTime display snapshot 1`] = `
<div>
  <div
    class="Image"
  >
    <span>
      <span
        class="LabelText"
      >
        Display
      </span>
      <span
        aria-label="Display"
        class="Image"
      >
        /media/examples/my-button.png
      </span>
    </span>
  </div>
</div>
`;

exports[`DataDateTime editing snapshot 1`] = `
<div>
  <div
    class="Image"
  >
    <label
      for="Editing"
    >
      <span
        class="LabelText"
      >
        Editing
      </span>
      <input
        alt="Start"
        class="Input"
        id="Editing"
        name="Editing"
        src="/media/examples/my-button.png"
        type="image"
      />
    </label>
  </div>
</div>
`;
